stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -219,10 +219,34 @@ end
 (*)(D::Diagonal, B::AbstractTriangular) =
     lmul!(D, copy_oftype(B, promote_op(*, eltype(B), eltype(D.diag))))
 
-(*)(A::AbstractMatrix, D::Diagonal) =
-    rmul!(copy_similar(A, promote_op(*, eltype(A), eltype(D.diag))), D)
-(*)(D::Diagonal, A::AbstractMatrix) =
-    lmul!(D, copy_similar(A, promote_op(*, eltype(A), eltype(D.diag))))
+# Using mul! instead of (l/r)mul! provides a performance boost by avoiding a copy.
+# This is not possible in general, but works if the elements are numbers
+# See #42321
+# Using runtime eltype checks instead of dispatch to avoid ambiguities
+function (*)(A::AbstractMatrix, D::Diagonal)
+    Tdest = promote_op(*, eltype(A), eltype(D.diag))
+    if eltype(A) <: Number && eltype(D) <: Number
+        Base.require_one_based_indexing(A)
```

Comment:
I don't see where one-based indexing matters here. There is no indexing going on, no `for i in 1:n` and alike.

Reply:
Diagonal’s diag vector is restricted to one-base indexing.

Reply:
Yes, but `rmul!` probably has it because it involves indexing which starts at 1. I assume the check is performed inside `mul!` and `rmul!` anyways.